packages/grid/_modules_/grid/hooks/root/useGridContainerProps.ts

```diff
@@ -121,14 +130,21 @@ export const useGridContainerProps = (
         return null;
       }
 
-      if (options.autoPageSize || options.autoHeight) {
-        // we don't need vertical virtualization in these 2 cases.
-        const viewportPageSize = options.autoHeight
-          ? rowsCount
-          : Math.floor(viewportSizes.height / rowHeight);
-        const requiredHeight = viewportPageSize * rowHeight + scrollState.scrollBarSize.x;
+      const requiredSize = rowsCount * rowHeight;
+      const diff = requiredSize - windowSizesRef.current.height;
+      // we activate virtualisation when we have more than 2 rows outside the viewport
+      const isVirtualized = diff > rowHeight * 2;
+
+      if (options.autoPageSize || options.autoHeight || !isVirtualized) {
+        // we don't need vertical virtualization in these cases.
+        const viewportPageSize =
+          options.autoHeight || !isVirtualized
+            ? rowsCount
+            : Math.floor(viewportSizes.height / rowHeight);
+        const requiredHeight = viewportPageSize * rowHeight;
 
         const indexes: GridContainerProps = {
+          isVirtualized,
```

Comment:
It looks like a leaky abstraction. Isn't virtualization supposed to only be a concern of `useGridVirtualRows.ts`? It seems that it's only computed here is it can be tested. This sounds like the opposite of an end-to-end test. Won't these tests be slow to update when we want to refactor the implementation?
```suggestion
```

Reply:
I wonder if always keeping the virtualization on could be simpler.

Reply:
You're not disabling when it suits. It is on by default when more than 2 rows are outside the viewport. So we can keep a natural experience when we have a small set of rows.

Reply:
I don't follow what you mean my natural experience. If there is no scrollbar, there is no virtulization, by definition. It still seems thar we have the DOM structure in place for virtulization when is Virtualize is false.

In any case, my main point of interpretation was about leaking a value to the public API that doesn't seem to solve any problem for developers, and to have it in place that doesn't seem related to virtualization (hence my question about the leaky abstraction).

Reply:
This value is in the grid state like all the other values that the grid needs. It is used in other places.
Why is it not related to virtualisation?

Reply:
> Why is it not related to virtualisation?

I meant that `isVirtualized` is computed in `useGridContainerProps.ts` while it's only used in `useGridVirtualRows.ts`. It seems that we could compute and consuming it in `useGridVirtualRows.ts`, as a locally scoped variable. This would also have the advantage of making it impossible to test in the end-to-end tests (so we can focus on the behavior instead).

Reply:
it's used in 2 places

https://github.com/mui-org/material-ui-x/pull/1253/files#diff-7904e8c986f79f8bbc1627613039ba9fc46cc3f000aef6b4883c56f4aff2311aR154

https://github.com/mui-org/material-ui-x/pull/1253/files#diff-b067e2d006d18d8c1214e487d80d2a267f8bc6e8b163055fb0c381317db7010fR138

Reply:
My bad, I didn't notice it it was read too.

I didn't look at the root issue, so I can only share my high-level perspective on the problem. It seems to go in the opposite direction to where we should aim:

1. It seems to melt `useGridVirtualRows` and `useGridContainerProps` together. Shouldn't we aim for more isolation?
2. It seems to create more branching rendering behavior. The more we have, the more challenging pinning, dynamic rows, tree data will be to handle. Shouldn't we aim for fixing the virtualization edge cases at their core?

In any case, the bugs are fixed, and we have a couple of new constraints that should help for the next time we iterate on the problem 👍  (the tests that depends on isVirtualized are unlikely to be very helpful in the future, but we can easily remove them if needed, hoping we won't lost the learnings from this iteration).